docs/best-practices.rst

```diff
@@ -0,0 +1,271 @@
+
+You should treat tasks in Airflow equivalent to transactions in a database. It implies that you should never produce
+incomplete results from your tasks. An example is not to produce incomplete data in ``HDFS`` or ``S3`` at the end of a task.
+
+Airflow retries a task if it fails. Thus, the tasks should produce the same outcome on every re-run.
+Some of the ways you can avoid producing a different result -
+
+* Do not use INSERT during a task re-run, an INSERT statement might lead to duplicate rows in your database.
+  Replace it with UPSERT.
+* Read and write in a specific partition. Never read the latest available data in a task. 
+  Someone may update the input data between re-runs, which results in different outputs. 
+  A better way is to read the input data from a specific partition. You can use ``execution_date`` as a partition. 
+  You should follow this partitioning method while writing data in S3/HDFS, as well.
+* The python datetime ``now()`` function gives the current datetime object. 
+  This function should never be used inside a task, especially to do the critical computation, as it leads to different outcomes on each run. 
+  It's fine to use it, for example, to generate a temporary log.
+
+.. tip::
+
+    You should define repetitive parameters such as ``connection_id`` or S3 paths in ``default_args`` rather than declaring them for each task.
+    The ``default_args`` help to avoid mistakes such as typographical errors.
+
+
+Deleting a task
+----------------
+
+Never delete a task from a DAG. In case of deletion, the historical information of the task disappears from the Airflow UI. 
+It is advised to create a new DAG in case the tasks need to be deleted.
+
+
+Communication
+--------------
+
+Airflow executes tasks of a DAG in different directories, which can even be present 
+on different servers in case you are using :doc:`Kubernetes executor <../executor/kubernetes>` or :doc:`Celery executor <../executor/celery>`. 
+Therefore, you should not store any file or config in the local filesystem — for example, a task that downloads the JAR file that the next task executes.
+
+Always use XCom to communicate small messages between tasks or S3/HDFS to communicate large messages/files.
+
+The tasks should also not store any authentication parameters such as passwords or token inside them. 
+Always use :ref:`Connections <concepts-connections>` to store data securely in Airflow backend and retrieve them using a unique connection id.
+
+
+Variables
+---------
+
+You should avoid usage of Variables outside an operator's execute() method or Jinja templates. Variables create a connection to metadata DB of Airflow to fetch the value.
+Airflow parses all the DAGs in the background at a specific period.
+The default period is set using ``processor_poll_interval`` config, which is by default 1 second. During parsing, Airflow creates a new connection to the metadata DB for each Variable.
+It can result in a lot of open connections.
```

Comment:
```
Airflow parses all the DAG files in a loop, trying to parse each file every ``processor_poll_interval`` seconds (default 1 second). During parsing, Airflow will open and close a new connection to the metadata DB for each Variable. Because of this using Variables at the "file-scope" of your DAG files should be avoided unless you have no choice.
```

(It should be only for each dag file, not for each Variable used. But I haven't tested this)

Reply:
From the airflow code, it seems like a new session is created per variable

Reply:
We configure things to use a session pool, or _should_.

But I have not tested this to see what actually happens in a long time :)

Reply:
Yes, you are right. I looked at the wrapper function only. The actual session created uses the pool.